### PR TITLE
Further exceptions tweaks

### PR DESCRIPTION
**Pull request by Majkl578**
Expands and refines #6743.

* All exceptions are made final (or abstract).
* Nameless constructors renamed to `new()`.
* Exceptions moved around to their appropriate places (Exception sub-namespace), except few that are to be handled later (i.e. Mapping, Query).
* Removed Exception suffixes.
* Also cleaned up root namespace by moving out non-namespaced exceptions.

To be documented after first round of review.

cc @greg0ire 

TODO:
* [ ] add method constructors for pre-3.0 exceptions

### CODE REVIEW

File: lib/Doctrine/ORM/Exception/PessimisticLockFailed.php [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+declare(strict_types=1);
+
+namespace Doctrine\ORM\Exception;
+
+final class PessimisticLockFailed extends \RuntimeException implements ORMException
+{
+    /**
+     * @return PessimisticLockFailed
+     */
+    public static function lockFailed()
```
Comment (carusogabriel) [inline lib/Doctrine/ORM/Exception/PessimisticLockFailed.php]: Return type maybe?

Comment (Majkl578) [reply lib/Doctrine/ORM/Exception/PessimisticLockFailed.php]: Probably not in this PR, not directly related. :)

File: lib/Doctrine/ORM/Exception/UnexpectedResult.php [hunk 2]
```diff
@@ -0,0 +1,12 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\ORM\Exception;
+
+/**
+ * Exception for a unexpected query result.
```
Comment (greg0ire) [inline lib/Doctrine/ORM/Exception/UnexpectedResult.php]: a => an

File: lib/Doctrine/ORM/Mapping/MappingException.php [hunk 3]
```diff
@@ -14,8 +14,10 @@
 
 /**
  * A MappingException indicates that something is wrong with the mapping setup.
+ *
+ * @todo Split into specific exceptions and change to interface.
```
Comment (greg0ire) [inline lib/Doctrine/ORM/Mapping/MappingException.php]: :+1:

**Review by greg0ire — COMMENTED** [review]
Beautiful! :sparkles:

File: lib/Doctrine/ORM/UnexpectedResultException.php [hunk 4]
```diff
@@ -8,6 +8,8 @@
 
 /**
  * Exception for a unexpected query result.
+ *
+ * @final
```
Comment (lcobucci) [inline lib/Doctrine/ORM/UnexpectedResultException.php]: What's the reason for having this as an annotation?

Comment (lcobucci) [reply lib/Doctrine/ORM/UnexpectedResultException.php]: Nevermind 😛

File: docs/en/reference/dql-doctrine-query-language.rst [hunk 5]
```diff
@@ -981,9 +981,9 @@ the Query class. Here they are:
    result is either a plain collection of objects (pure) or an array
    where the objects are nested in the result rows (mixed).
 -  ``Query#getSingleResult()``: Retrieves a single object. If the
-   result contains more than one object, an ``NonUniqueResultException``
-   is thrown. If the result contains no objects, an ``NoResultException``
-   is thrown. The pure/mixed distinction does not apply.
+   result contains more than one object, an ``NonUniqueResult``
```
Comment (malukenho) [inline docs/en/reference/dql-doctrine-query-language.rst]: an => a

File: docs/en/reference/dql-doctrine-query-language.rst [hunk 6]
```diff
@@ -981,9 +981,9 @@ the Query class. Here they are:
...
+   is thrown. If the result contains no objects, an ``NoResult``
```
Comment (malukenho) [inline docs/en/reference/dql-doctrine-query-language.rst]: an => a

**Pull request closed without merging**